src/Resources/contao/dca/tl_user.php

```diff
@@ -985,4 +1019,155 @@ public function toggleVisibility($intId, $blnVisible, DataContainer $dc=null)
 
 		$objVersions->create();
 	}
+
+	/**
+	 * Builds the palette for 2FA dynamically.
+	 *
+	 * @param DataContainer $dc
+	 */
+	public function build2faPalette(DataContainer $dc)
+	{
+		$activeRecord = UserModel::findById((int) $dc->id);
+
+		if ($activeRecord === null)
+		{
+			return;
+		}
+
+		foreach ($GLOBALS['TL_DCA']['tl_user']['palettes'] as $palette => $v)
+		{
+			if ($palette === '__selector__')
+			{
+				continue;
+			}
+
+			// Don't show 2FA options, if it is not the logged in user itself and 2FA is disabled.
+			if ($dc->id !== $this->User->id && (int) $activeRecord->use2fa !== '1')
+			{
+				continue;
+			}
+
+			$GLOBALS['TL_DCA']['tl_user']['palettes'][$palette] = str_replace
+			(
+				',password',
+				',password;{2fa_legend},use2fa',
+				$GLOBALS['TL_DCA']['tl_user']['palettes'][$palette]
+			);
+		}
+
+		// Only add subpalette if the logged in user is editing his own record
+		if ((int) $dc->id === (int) $this->User->id)
+		{
+			// extend selector
+			$GLOBALS['TL_DCA']['tl_user']['palettes']['__selector__'][] = 'use2fa';
+			$GLOBALS['TL_DCA']['tl_user']['subpalettes']['use2fa'] = '2faQrCode';
+
+			if ($this->User->confirmed2fa !== '1')
+			{
+				$GLOBALS['TL_DCA']['tl_user']['subpalettes']['use2fa'] .= ',confirmed2fa';
+			}
+		}
+	}
+
+	/**
+	 * Generates a 2FA secret if not present.
+	 *
+	 * @param DataContainer $dc
+	 *
+	 * @return string
+	 */
+	protected function generate2faSecret(DataContainer $dc)
+	{
+		$secret = $dc->activeRecord->secret;
+
+		if ($secret === null)
+		{
+			// Generate 1024 bit secret
+			$secret = random_bytes(128);
+
+			$this->Database->prepare("UPDATE tl_user SET secret=? WHERE id=?")
```

Comment:
Shouldn't that be something like this?

```php
$user = BackendUser::getInstance();
$user->secret = random_bytes(128);
$user->save();
```

Reply:
Yep. Can't see the forest for the trees.